src/VisualStudio/Core/Def/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -431,24 +432,38 @@ public async ValueTask RefreshFileAsync(CancellationToken cancellationToken)
 
         private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
         {
-            var oldProject = e.OldSolution.GetProject(_documentIdentity.DocumentId.ProjectId);
-            var newProject = e.NewSolution.GetProject(_documentIdentity.DocumentId.ProjectId);
+            var projectId = _documentIdentity.DocumentId.ProjectId;
+
+            // Trivial check.  see if the SG version of these projects changed.  If so, we definitely want to update
+            // this generated file.
+            if (e.OldSolution.GetSourceGeneratorExecutionVersion(projectId) !=
+                e.NewSolution.GetSourceGeneratorExecutionVersion(projectId))
+            {
+                _batchingWorkQueue.AddWork();
+                return;
+            }
+
+            var oldProject = e.OldSolution.GetProject(projectId);
+            var newProject = e.NewSolution.GetProject(projectId);
 
             if (oldProject != null && newProject != null)
             {
                 // We'll start this work asynchronously to figure out if we need to change; if the file is closed the cancellationToken
                 // is triggered and this will no-op.
                 var asyncToken = _fileManager._listener.BeginAsyncOperation($"{nameof(OpenSourceGeneratedFile)}.{nameof(OnWorkspaceChanged)}");
+                CheckDependentVersionsAsync().CompletesAsyncOperation(asyncToken);
+            }
 
-                var cancellationToken = _cancellationTokenSource.Token;
-                Task.Run(async () =>
+            async Task CheckDependentVersionsAsync()
```

Comment:
> can you clarify GetSourceGeneratorExecutionVersion? Is it possible for that version to not change, but the generated file needs to be updated?

Yes, 100%.  

In *automatic* mode, nothing is really updating the version ever  (unless you hit the explicit 'rerun generaotr' button).  So the version just stays the same.  But then we may rerun generators when we see any other change has happened (like a code edit).

In *balanced* mode, the compiler-trackers moves out of automatically running generators, and so the 'version' flag is a way to 'kick it' to get it to run, independent of all the other events that will *not* cause rerunning.

Importantly, the version also acts as a way to ensure that the states of the system actually *change* when they would otherwise not.  Consider, for example, that we want to rerun generators on 'save'.  Well, saving doesn't actually cause any normal state/checksums to change.  After all, the file content is the same before/after.  So if we want a system that actually reruns (including making sure an externally sync'ed process like OOP reruns), then we need something to eb part of our state/checksum to indicate that.  The SourceGenExecutionVersion acts as that, allowing the external host to effectively mix-in its state into our own snapshot computation/syncing.

LMK if i can explain this better or if you need more info here.